src/Chapter3.hs

```diff
@@ -480,6 +520,50 @@ After defining the city, implement the following functions:
    and at least 10 living __people__ inside in all houses of the city totally.
 -}
 
+data Castle = Castle 
+  { castleName :: String
+  , hasWalls :: Bool
+  } deriving (Show)
+
+makeCastle :: String -> Bool -> Castle
+makeCastle n withWalls = Castle { castleName = n, hasWalls = True }
+
+data PersonCount = One | Two | Three | Four deriving (Show)
+data House = House PersonCount deriving (Show)
+
+houseCount :: House -> Int
+houseCount (House One) = 1
+houseCount (House Two) = 2
+houseCount (House Three) = 3
+houseCount (House Four) = 4
+
+data Building = Church | Library deriving (Show)
+
+data City = City 
+  { cityCastle :: Maybe Castle
```

Comment:
This is the correct solution :+1: But I feel like introducing a custom sum type will be more convenient here.

First, we need to decide with the castle. Let's look at the task:

> Wall, but only if the city has a castle
> -- Optional castle with a name (as 'String')

That means that walls are very connected to the castle, and they could not exist separately. So we have three options for the castle:

1. Castle with a name
2. Castle with the name + walls
3. No castle

This looks like a sum type!:

```haskell
data Castle 
    = Castle String
    | WallCastle String
    | None
```

Note that your idea with castle suits here perfectly as well. But you need to define such type separately (as you did):

```haskell
data Castle = Castle 
    { castleName :: String
    , castleWall :: Bool 
    }
```

And in the city, we will have `Maybe Castle`. More nested data types and a bit less convenient to work with. Creating new data types in Haskell is extremely cheap :slightly_smiling_face:

Reply:
> More nested data types and a bit less convenient to work
I agree with this point. 
Implementation looks much cleaner now. Thanks a lot for your suggestion 😄